### PR TITLE
Enable bootstrapping non-build-machine targets

### PR DESCRIPTION
**Pull request by alexcrichton**
These commits add support to the rustbuild build system to compile non-build-system compilers. For example this will allow creating an ARM compiler on a x86 system. The high level way this works is:
- The only compiler ever run is the build target compiler. No other compiler is assumed to be runnable.
- As a result, all output artifacts come from the build compiler.
- The libs for the stageN cross-compiled compiler will be linked into place after the build target builds them.

This will break the assumption that a compiler never links to anything it didn't itself produce, but it retains the assumption that a compiler only ever links to anything built in the same stage. I believe this means that the stage1 cross-compiled compilers will still be usable.

I tested this by creating an `arm-unknown-linux-gnueabihf` compiler. So far the linking ended up all working OK (including LLVM being cross compiled), but I haven't been able to run it yet (in QEMU for a raspberry pi). I think that's because my system linker is messed up or something like that (some newer option is assumed when it's not actually there).

Overall, though, this means that rustbuild can compile an arm-unknown-linux-gnueabihf compiler. Ideally we could even start shipping nightlies based on this at some point!

### CODE REVIEW

File: src/bootstrap/build/native.rs [hunk 1]
```diff
@@ -115,6 +115,11 @@ pub fn compiler_rt(build: &Build, target: &str) {
     let mode = if build.config.rust_optimize {"Release"} else {"Debug"};
     let (dir, build_target, libname) = if target.contains("linux") {
         let os = if target.contains("android") {"-android"} else {""};
+        let arch = if arch == "arm" && target.contains("eabihf") {
```
Comment (japaric) [inline src/bootstrap/build/native.rs]: does this cover the `armv7-unknown-linux-gnueabihf` target?

**Comment by japaric** [comment]
I've tested this PR by cross compiling a compiler to arm-unknown-linux-gnueabihf. And, then testing that cross compiled rustc/std on an ARMv7 board by running some smoke tests. The smoke tests consisted of: building cargo, cargo testing hyper and cargo testing clap and all of them succeeded. (NOTE: I needed cargo on the ARM target, so I used an unofficial cargo binary from RustBuild. The host was an x86_64 machine running Ubuntu 15.10. The target was an ARMv7 machine running hard-float Ubuntu 15.10)

I found two (minor) issues. @alexcrichton  is already aware of them, but to reiterate:

- We use the flags produced by the host `llvm-config --cxxflags` to cross compile rustc_llvm. Among these flags there is a `-m64` flag that's not recognized by the ARM cross g++ compiler. This flag needs to be filtered out before passing the flags to g++.
- The cross compiled `libcompiler-rt.a` doesn't get copied to the `build/arm-unknown-linux-gnueabihf/stageN` directories.

File: src/bootstrap/build/native.rs (hunk 1 above)
Comment (alexcrichton) [inline src/bootstrap/build/native.rs]: I... doubt it will!

**Comment by alexcrichton** [comment]
Thanks for testing this out @japaric! I've amended the relevant commits and those issues should all be fixed.

**Comment by joerg-krause** [comment]
How do you exactly use this feature? Can you give an example on howto build for the ARM target you mentioned?

**Comment by japaric** [comment]
> How do you exactly use this feature? Can you give an example on howto build for the ARM target you mentioned?

Pass `--host=$triple` and `--enable-rustbuild` to the `configure` script before calling `make`. Something like this:

```
mkdir build
cd build
../configure --enable-rustbuild --host=arm-unknown-linux-gnueabihf
make
```

**Pull request merged**